### PR TITLE
Support sub project build without parent Cargo.toml.

### PR DESCRIPTION
**Pull request by avkviring**

### CODE REVIEW

**Comment by avkviring** [comment]
Project Structure:
- root
-- Client/Cargo.toml
-- Common/Cargo.toml
-- Server/Cargo.toml


cross build --manifest-path clients/Cargo.toml --target armv7-linux-androideabi --release

File: src/cli.rs [hunk 1]
```diff
@@ -13,18 +13,24 @@ pub struct Args {
...
     let mut target = None;
+    let mut project_dir: Option<String> = None;
     let mut target_dir = None;
     let mut sc = None;
     let mut all: Vec<String> = Vec::new();
 
     {
         let mut args = env::args().skip(1);
         while let Some(arg) = args.next() {
+            if all.last().unwrap_or(&"".to_string()) == "--manifest-path" {
```
Comment (reitermarkus) [inline src/cli.rs:30]: Why `all.last()` when we are using an iterator? Use the same pattern as for the other flags below.

File: src/cli.rs [hunk 2]
```diff
@@ -13,18 +13,24 @@ pub struct Args {
     pub target: Option<Target>,
     pub target_dir: Option<PathBuf>,
     pub docker_in_docker: bool,
+    pub project_dir: Option<String>,
 }
 
 pub fn parse(target_list: &TargetList) -> Args {
     let mut channel = None;
     let mut target = None;
+    let mut project_dir: Option<String> = None;
```
Comment (reitermarkus) [inline src/cli.rs:22]: Use a `PathBuf` to store this.

File: src/cli.rs [hunk 3]
```diff
@@ -13,18 +13,24 @@ pub struct Args {
...
     let mut target_dir = None;
     let mut sc = None;
     let mut all: Vec<String> = Vec::new();
 
     {
         let mut args = env::args().skip(1);
         while let Some(arg) = args.next() {
+            if all.last().unwrap_or(&"".to_string()) == "--manifest-path" {
+                project_dir = Option::Some(format!("{}/{}", env::current_dir().expect("").to_str().unwrap(), arg.to_string()));
```
Comment (reitermarkus) [inline src/cli.rs:31]: Use `Path::join` instead of `format!`.

**Comment by reitermarkus** [comment]
> This command don't work:
> `cross build --manifest-path clients/Cargo.toml --target armv7-linux-androideabi --release`

Can you pose the output of the failed command?

File: src/cli.rs (hunk 1 above)
Comment (avkviring) [reply src/cli.rs]: fix

File: src/cli.rs (hunk 2 above)
Comment (avkviring) [reply src/cli.rs]: fix

File: src/cli.rs (hunk 3 above)
Comment (avkviring) [reply src/cli.rs]: fix

**Comment by avkviring** [comment]
> > This command don't work:
> > `cross build --manifest-path clients/Cargo.toml --target armv7-linux-androideabi --release`
> 
> Can you post the output of the failed command?

``

error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-Wl,--allow-multiple-definition" "-march=armv7-a" "-L" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/dep
s/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.0.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.1.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideab
i/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.10.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.11.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-l
inux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.12.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.13.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust
/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.14.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.15.rcgu.o" "/Users/kviring/Documents/hive/rel
ay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.2.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.3.rcgu.o" "/Users/kviring/Docum
ents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.4.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.5.rcgu.o" "/Users/
kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.6.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.7.rcg
u.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167u73-cgu.8.rcgu.o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/cheetah_relay_client.cheetah_relay_client.ev167
u73-cgu.9.rcgu.o" "-o" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/libcheetah_relay_client.so" "-Wl,--version-script=/var/folders/qq/781_h86n6t9fbt6jn2nbccnw0000gn/T/rustcCxezox/list" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-an
droideabi/release/deps/cheetah_relay_client.1wx5ir8dkg0p924p.rcgu.o" "-Wl,--gc-sections" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps" "-L" "/Users/kviring/Documents/hive/relay/clients/Rust/target/r
elease/deps" "-L" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib" "-Wl,-Bstatic" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/libcheetah_relay_common-a80844b66149cc23.rlib" "/Users/kviring/Documents/hi
ve/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/libmio-fae17cede4d64ee4.rlib" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/liblibc-3c7d570ede60bd7b.rlib" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/rel
ease/deps/libwidestring-95ea1097bac52adc.rlib" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/liblog-87ef413bac3b1073.rlib" "/Users/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/libcfg_if-abb5a9b7f09a3c74.rlib" "/Users
/kviring/Documents/hive/relay/clients/Rust/target/armv7-linux-androideabi/release/deps/liblazy_static-9083aaad39adfd41.rlib" "-Wl,--start-group" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libstd-609ec648101596d0.rlib" "/Users/kviring/.rustup/tool
chains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libpanic_unwind-d121513aa656fa25.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libhashbrown-76c34cca3b4ae345.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple
-darwin/lib/rustlib/armv7-linux-androideabi/lib/librustc_std_workspace_alloc-406df22057718a84.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libbacktrace-dd7a96a0b03e0281.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/r
ustlib/armv7-linux-androideabi/lib/libbacktrace_sys-d962076d47e2542f.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/librustc_demangle-da7738c52a2d7de1.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-a
ndroideabi/lib/libunwind-a5b2db70706365f5.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libcfg_if-b6c531a607a8b8dc.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/liblibc-cc4cd396b234
de52.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/liballoc-af413bd74fcfa4f6.rlib" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/librustc_std_workspace_core-b14bdcbf8defb6a1.rlib" "/Users
/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libcore-9f486e96edfae6f5.rlib" "-Wl,--end-group" "/Users/kviring/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/armv7-linux-androideabi/lib/libcompiler_builtins-9b2796896fdb0eb1.rlib" "-Wl,-Bdynamic"
 "-ldl" "-llog" "-lgcc" "-lc" "-lm" "-shared"
  = note: ld: unknown option: --as-needed
          clang: error: linker command failed with exit code 1 (use -v to see invocation)

File: src/cli.rs [hunk 4]
```diff
@@ -13,19 +13,26 @@ pub struct Args {
...
     let mut all: Vec<String> = Vec::new();
-
+    
     {
         let mut args = env::args().skip(1);
         while let Some(arg) = args.next() {
-            if let ("+", ch) = arg.split_at(1) {
+            if arg == "--manifest-path" {
+                all.push(arg);
+                let path = args.next().expect("Missing argument in --manifest-path");
```
Comment (reitermarkus) [inline src/cli.rs:32]: Use `if let Some(..)` for this and also add the case for `--manifest-path=`.

File: src/cargo.rs [hunk 5]
```diff
@@ -26,7 +26,7 @@ impl Subcommand {
             _ => true,
         }
     }
-
+
```
Comment (reitermarkus) [inline src/cargo.rs:29]: Please fix the trailing whitespace everywhere.

File: src/cli.rs [hunk 6]
```diff
@@ -13,19 +13,26 @@ pub struct Args {
...
+                project_dir = Option::Some(env::current_dir().expect("").join(PathBuf::from(&path)));
```
Comment (reitermarkus) [inline src/cli.rs:33]: Add a message to `expect`.

File: src/cli.rs (hunk 4 above)
Comment (avkviring) [reply src/cli.rs:32]: > > This command don't work:
> > `cross build --manifest-path clients/Cargo.toml --target armv7-linux-androideabi --release`
> 
> Can you post the output of the failed command?

It is fallback to host cargo and fail in linking

File: src/cargo.rs (hunk 5 above)
Comment (avkviring) [reply src/cargo.rs:29]: fix

File: src/cli.rs (hunk 4 above)
Comment (avkviring) [reply src/cli.rs:32]: I don't understand :-( I am newbies in rust, please help me.

Comment (reitermarkus) [reply src/cli.rs:32]: Look at the code below, there it is used in the same way.

Comment (reitermarkus) [reply src/cli.rs:32]:
```rust
if let Some(path) = args.next() {
```

Comment (avkviring) [reply src/cli.rs]: ok, fix

File: src/cargo.rs (hunk 5 above)
Comment (reitermarkus) [reply src/cargo.rs:29]: The whitespace is still there, please remove it.

File: src/cli.rs [hunk 7]
```diff
@@ -13,19 +13,27 @@ pub struct Args {
...
     let mut target_dir = None;
     let mut sc = None;
     let mut all: Vec<String> = Vec::new();
-
+    
     {
         let mut args = env::args().skip(1);
         while let Some(arg) = args.next() {
-            if let ("+", ch) = arg.split_at(1) {
+            if arg == "--manifest-path" {
```
Comment (reitermarkus) [inline src/cli.rs:30]: Still needs the case for `--manifest-path=`.

**Comment by reitermarkus** [comment]
@avkviring, any update on this?

**Comment by avkviring** [comment]
> @avkviring, any update on this?
Hi, I have vacation and don't have computer now :-(

**Comment by wngr** [comment]
I'd be interested in taking this over, if there is no disagreement.

**Comment by avkviring** [comment]
Hi, it is very good news :-)

**Pull request closed without merging**